### PR TITLE
Check for valid renamed usernames

### LINKED ISSUE #2066: Bad, and possibly unsafe usernames
If you create a user, and rename it to one of the following:

- `%2f.` It redirects you to `https://.` when someone tries to view your profile
- `%2f..` It redirects you to `https://..` when someone tries to view your profile
- `%2f*` 404 when someone tries to view your profile
- `%00` It redirects you to a blank page when someone tries to view your profile

These have been tested in opera and chrome.

### PR DESCRIPTION
**Pull request by ethantkoenig**
Fixes #2066 and #2061.

### CODE REVIEW

**Comment by lafriks** [comment]
LGTM

**Comment by lafriks** [comment]
May be add integration test?

**Comment by lunny** [comment]
LGTM otherwise @lafriks 's comment.

**Comment by ethantkoenig** [comment]
@lafriks @lunny Integration test added :smile:

**Comment by lafriks** [comment]
Awsome 👍

File: integrations/user_test.go [hunk 1]
```diff
@@ -18,3 +21,27 @@ func TestViewUser(t *testing.T) {
...
+
+	session := loginUser(t, "user2")
+	req := NewRequest(t, "GET", "/user/settings")
+	resp := session.MakeRequest(t, req)
+	assert.EqualValues(t, http.StatusOK, resp.HeaderCode)
+
+	htmlDoc := NewHTMLParser(t, resp.Body)
+	req = NewRequestWithValues(t, "POST", "/user/settings", map[string]string{
+		"_csrf": htmlDoc.GetCSRF(),
+		"name":  "%2f*", // not valid
```
Comment (bkcsoft) [inline integrations/user_test.go]: Could you add more tests here? The original issue lists a bunch of cases that we should guard against 🤔 

```go
tests := []string{ ... }
for test := range tests {
  t.Logf("testing %q", test)
  req = NewRequest...("POST" ...)
  [...]
  models.AssertNot...
}
```
( ☝️ Table-driven testing 🙂 https://blog.golang.org/subtests )

**Review by bkcsoft — CHANGES_REQUESTED** [review]

File: integrations/user_test.go (hunk 1 above)
Comment (lafriks) [reply integrations/user_test.go]: It would be also nice to have valid user name change test

Comment (ethantkoenig) [reply integrations/user_test.go]: Added all invalid cases from #2066, and a reserved username test, and a valid username test.

File: integrations/user_test.go [hunk 2]
```diff
@@ -22,26 +22,97 @@ func TestViewUser(t *testing.T) {
...
+		assert.EqualValues(t, http.StatusFound, resp.HeaderCode)
+
+		req = NewRequest(t, "GET", "/user/settings")
+		resp = session.MakeRequest(t, req)
+		assert.EqualValues(t, http.StatusOK, resp.HeaderCode)
+		htmlDoc = NewHTMLParser(t, resp.Body)
+		assert.Contains(t,
+			htmlDoc.doc.Find(".ui.negative.message").Text(),
+			i18n.Tr("en", "user.newName_reserved"),
+		)
```
Comment (bkcsoft) [inline integrations/user_test.go:115]: now you have 3 identical loops, have them be a common function like so `testUpdateUser(t *testing.T, username, email string, code http.StatusCode)` 🙂

Comment (ethantkoenig) [reply integrations/user_test.go:115]: @bkcsoft They're not identical. A lot of the setup code (e.g. getting the CSRF token) is shared, but once you make the `POST /user/settings` request, they have little in common (different status codes, may or may not have to follow a redirect, may or may not check for whether rename succeeded, have to check for different error messages).

The boilerplate setup code is also shared with tests from other files, so IMO the best solution is to add helper functions for things like getting a CSRF token.

**Comment by lafriks** [comment]
Please also add test case for username with space so that #2061 can be also closed

**Comment by ethantkoenig** [comment]
@lafriks Done

**Comment by lunny** [comment]
need @bkcsoft approval.

File: modules/auth/user_form.go [hunk 3]
```diff
@@ -100,7 +100,7 @@ func (f *SignInForm) Validate(ctx *macaron.Context, errs binding.Errors) binding
 
 // UpdateProfileForm form for updating profile
 type UpdateProfileForm struct {
-	Name             string `binding:"OmitEmpty;MaxSize(35)"`
+	Name             string `binding:"OmitEmpty;AlphaDashDot;MaxSize(35)"`
```
Comment (lafriks) [inline modules/auth/user_form.go]: I think it is also needs to be `Required`

**Comment by lafriks** [comment]
Related https://github.com/gogits/gogs/commit/e6dbfd918c8d68c29c01bf2617321b037393ada1
https://www.sourceclear.com/registry/security/cross-site-scripting-xss-/go/sid-4456

**Comment by lafriks** [comment]
@bkcsoft need your approval

**Review by bkcsoft — APPROVED** [review]

**Pull request merged**